fix(runtime): fix multiprocess runtime initialization bug

amha-kindu opened the pull request:
<!--

CodeRabbit will automatically analyze your PR and provide:
- Code review comments
- Security analysis
- Performance insights
No action needed in this section - let AI do the heavy lifting!

@coderabbitai
-->

clairbee commented:
This does prevent another process from assuming that the sandbox is already created when it's not. But it doesn't make the other process wait for it to be created. As a result, more than one copy will be created at once overwriting each other, potentially failing as well (e.g. conda may complain that the environment already exists, if the other process manages to create the directory after PartCAD checks whether the folder exists and before conda checks whether the environment is already there). A file next to the sandbox directory (e.g. "<sandbox-dir>.lck") should be used as a filesystem lock (instead of "threading.lock()").

amha-kindu commented:
> This does prevent another process from assuming that the sandbox is already created when it's not. But it doesn't make the other process wait for it to be created. As a result, more than one copy will be created at once overwriting each other, potentially failing as well (e.g. conda may complain that the environment already exists, if the other process manages to create the directory after PartCAD checks whether the folder exists and before conda checks whether the environment is already there). A file next to the sandbox directory (e.g. ".lck") should be used as a filesystem lock (instead of "threading.lock()").

I didn't encounter that particular case during testing but I understand your point. I'll introduce a filesystem lock.

clairbee commented:
> I didn't encounter that particular case during testing but I understand your point. I'll introduce a filesystem lock.

It might be even a bigger problem.
See https://github.com/partcad/partcad/actions/runs/13231899963/job/36930306439?pr=281
Apparently, a filesystem lock is needed for all conda operations.
It seems like we can't create two different conda environments at the same time.

Please, note, such a conda specific lock needs to be in the conda specific class (runtime_python_conda.py).
There still needs to be a sandbox specific lock in the sanbox class (runtime.py).
It will be useful not only for conda sandboxes, but for docker or qemu as well.

amha-kindu commented:
> > I didn't encounter that particular case during testing but I understand your point. I'll introduce a filesystem lock.
> 
> It might be even a bigger problem. See https://github.com/partcad/partcad/actions/runs/13231899963/job/36930306439?pr=281 Apparently, a filesystem lock is needed for all conda operations. It seems like we can't create two different conda environments at the same time.
> 
> Please, note, such a conda specific lock needs to be in the conda specific class (runtime_python_conda.py). There still needs to be a sandbox specific lock in the sanbox class (runtime.py). It will be useful not only for conda sandboxes, but for docker or qemu as well.

I have already modified python_runtime.py to use specific filesystem locks in VenvLock & AsyncVenvLock(which will lock during runtime operations) and introduced FileLock in PythonRuntime for initialization. However, I wasn't aware that we can't create two different conda environments at the same time so the locks were specific to each runtime type + version

partcad-admin commented on partcad/src/partcad/runtime_python.py:
```diff
@@ -89,6 +119,25 @@ def __init__(self, ctx, sandbox, version=None):
         if platform.system() == "Windows":
             self.pip_install_flags += ["--no-warn-script-location"]
 
+    @property
```
It's a nice try to make an improvement but there is more logic inside "ensure_*" method than it may seem initially.
Please, skip this optimization attempt and maintain all the same calls to "ensure_*" methods.

amha-kindu replied:
It's taken care of

partcad-admin commented on partcad/src/partcad/runtime_python.py:
```diff
@@ -124,16 +172,16 @@ async def once_async(self):
         return self.run_onced(cmd, stdin=stdin, cwd=cwd, session=session)
 
     def run_onced(self, cmd, stdin="", cwd=None, session=None, path=None):
-        if session and session["dirty"]:
```
What's the reason for this change?

clairbee commented on partcad/src/partcad/runtime_python.py:
```diff
@@ -150,51 +198,52 @@ def run_onced(self, cmd, stdin="", cwd=None, session=None, path=None):
         return await self.run_async_onced(cmd, stdin=stdin, cwd=cwd, session=session)
 
     async def run_async_onced(self, cmd, stdin="", cwd=None, session=None, path=None):
-        if session and session["dirty"]:
```
Again the same change.
Not sure what is it for?
Seems like the only side effect of this change is that sometimes the "AsyncVenvLock()" gets created, but not clear why.

clairbee replied:
Oh, I think I get it now. Hard to see the indentation in this web ui

clairbee commented:
Please, ignore the coverage requirements for this one.
